core/iwasm/libraries/libc-wasi/libc_wasi.cmake

```diff
@@ -5,6 +5,12 @@ set (LIBC_WASI_DIR ${CMAKE_CURRENT_LIST_DIR})
 
 add_definitions (-DWASM_ENABLE_LIBC_WASI=1)
 
+if (WAMR_BUILD_ENABLE_LIBC_WASI_SYNC_FLAGS EQUAL 1)
+    add_definitions (-DWASM_ENABLE_LIBC_WASI_SYNC_FLAGS=1)
+else()
+    add_definitions (-DWASM_ENABLE_LIBC_WASI_SYNC_FLAGS=0)
```

Comment:
No need to add the else branch if the macro is set to 0 by default in ssp_config.h.

Reply:
do we want to have a user-settable option at all?
if O_RSYNC is known broken on Linux, isn't it more straightforward to use `#if defined(__Linux__)`?

Reply:
I'm not sure what guarantees other POSIX-like OS's make about O_RSYNC so `#if defined(__linux__)` probably wouldn't be enough. 

If the user knows the host OS they're going to be running WAMR on supports all these sync flags, they can enable support. So I'd still say it's worth having a user option for that use case.

Reply:
do you mean all CONFIG_HAS_ flags should be a user option?

Reply:
no, just this flag. The other `CONFIG_HAS_` flags can be deduced from the platform which is being compiled for and would cause a compile error if they were incorrectly set. afaik there is no such way to deduce whether a platform supports `O_RSYNC` etc. via macros - if there were, we could get rid of the user option.

N.B. the existence of `O_RSYNC` flag _doesn't imply_ it's supported.

Reply:
IMO, it's reasonable to assume the existence of O_RSYNC implies some support of it.
we can add platform ifdefs for known exceptions.

Reply:
@yamt do you mean that if macro O_RSYNC is defined, then the fd sync flags should be supported? If that, maybe we can just add simple control in posix.c. But from @zoraaver's description, it's not yet clear whether these flags should be added to the standard or not, see https://github.com/WebAssembly/wasi-filesystem/issues/98, so we add the building option to let developer to choose.

Reply:
> @yamt do you mean that if macro O_RSYNC is defined, then the fd sync flags should be supported? If that, maybe we can just add simple control in posix.c.

i meant something like
```
#if defined(O_RSYNC) && !defined(__linux__)
#define HAVE_O_RSYNC
#endif
```

> But from @zoraaver's description, it's not yet clear whether these flags should be added to the standard or not, see [WebAssembly/wasi-filesystem#98](https://github.com/WebAssembly/wasi-filesystem/issues/98), so we add the building option to let developer to choose.

these flags are defined in preview1.
the mentioned discussion is about preview2.

Reply:
> these flags are defined in preview1.

Yup, sorry I got a bit mixed up. These flags are already defined in preview1 like you mentioned. But I _still_ think we shouldn't support them unconditionally. A user option could be a bit more reliable than deducing support from the OS + existence of the sync flag. But I don't have a strong opinion on it either way tbh, as long as we don't support them by default since that's misleading for users whose programs might depend on the data guarantees offered by `R_SYNC` etc.

> But from @zoraaver's description

That's my bad, I'll edit the description - I didn't realize that issue was referring to preview2 and not preview1 (which appears to be finalized).

Reply:
the i/o integrity guarantees are complex and involves many components; OS, filesystem, disk, battery, etc.
cf. https://www.postgresql.org/docs/current/wal-reliability.html

having said that, my impression is that it isn't wasm runtime (or user who configures the runtime) who is responsible to determine what integrity the whole system provides.
IMO, where the underlying platform is supposed to provide the functionality, it's enough for us to pass through the requests to the underlying platform.
how do you think?

Reply:
@yamt do you mean to define the CONFIG_HAS_O_xxx macros in ssp_config.h, and use them to control the code in posix.c? And don't add WAMR_BUILD_ENABLE_LIBC_WASI_SYNC_FLAGS and CONFIG_HAS_LIBC_WASI_SYNC_FLAGS?

For example, in ssp_config.h, add:
```C
#if defined(O_DSYNC) && !defined(__linux__)
#define CONFIG_HAS_O_DSYNC
#endif

#if defined(O_RSYNC) && !defined(__linux__)
#define CONFIG_HAS_O_RSYNC
#endif

#if defined(O_SYNC) && !defined(__linux__)
#define CONFIG_HAS_O_SYNC
#endif
```
in posix.c:
```C
    if (fs_flags & __WASI_FDFLAG_DSYNC != 0) {
#ifdef CONFIG_HAS_O_DSYNC
        noflags |= O_DSYNC;
        needed_inheriting |= __WASI_RIGHT_FD_DATASYNC;
#else
        return __WASI_ENOTSUP;
#endif
    }

    if (fs_flags & __WASI_FDFLAG_RSYNC != 0) {
#ifdef CONFIG_HAS_O_RSYNC
        noflags |= O_RSYNC;
        needed_inheriting |= __WASI_RIGHT_FD_SYNC;
#else
        return __WASI_ENOTSUP;
#endif
    }

    if (fs_flags & __WASI_FDFLAG_SYNC != 0) {
#ifdef CONFIG_HAS_O_SYNC
        noflags |= O_SYNC;
        needed_inheriting |= __WASI_RIGHT_FD_SYNC;
#else
        return __WASI_ENOTSUP;
#endif
    }
```
It is good to me if it can resolve the issue to run the rust WASI tests.

Reply:
> @yamt do you mean to define the CONFIG_HAS_O_xxx macros in ssp_config.h, and use them to control the code in posix.c? And don't add WAMR_BUILD_ENABLE_LIBC_WASI_SYNC_FLAGS and CONFIG_HAS_LIBC_WASI_SYNC_FLAGS?

yes.

> It is good to me if it can resolve the issue to run the rust WASI tests.

it doesn't solve the test issue.

my suggestion for the test issue: https://github.com/WebAssembly/wasi-testsuite/pull/91

Reply:
>IMO, where the underlying platform is supposed to provide the functionality, it's enough for us to pass through the requests to the underlying platform.

thanks, that makes sense actually.

I've implemented the suggestion to provide support based on the definition of these sync flags.